community/index/src/main/java/org/neo4j/index/internal/gbptree/TreeStatePair.java

```diff
@@ -118,11 +118,18 @@ else if ( stateA.isValid() && stateB.isValid() )
         {
             // return newest
 
-            // compare unstable generations of A/B and include sanity check for stable generations
-            // such that there cannot be a state S compared to other state O where
+            // compare unstable generations of A/B, if equal, compare clean flag (clean is newer than dirty)
+            // and include sanity check for stable generations such that there cannot be a state S compared
+            // to other state O where
             // S.unstableGeneration > O.unstableGeneration AND S.stableGeneration < O.stableGeneration
 
-            if ( stateA.stableGeneration() >= stateB.stableGeneration() &&
+            if ( stateA.stableGeneration() == stateB.stableGeneration() &&
+                    stateA.unstableGeneration() == stateB.unstableGeneration() &&
+                    stateA.isClean() != stateB.isClean() )
+            {
+                return Optional.of( stateA.isClean() ? stateA : stateB );
```

Comment:
For other illegal states like that, for example `stateA.stable > stateB.stable` + `stateA.unstable < stateB.unstable`, we just return `Optional.empty`. Error handling is different depending on if we ask for newest or oldest tree state and is handled elsewhere.